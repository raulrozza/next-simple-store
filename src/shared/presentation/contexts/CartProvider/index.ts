export { default as useAddToCart } from './hooks/useAddToCart';
