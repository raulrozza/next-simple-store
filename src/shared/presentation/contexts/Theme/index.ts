export { default as useTheme } from './hooks/useTheme';
export { default as Theme } from './implementations/StyledComponentTheme';
