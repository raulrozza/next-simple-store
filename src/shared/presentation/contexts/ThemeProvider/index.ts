export { default as useTheme } from './hooks/useTheme';
export { default as ThemeProvider } from './implementations/StyledComponentTheme';
