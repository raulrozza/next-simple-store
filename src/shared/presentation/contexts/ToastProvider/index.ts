export { useToastProvider } from './context';
export { default as ToastProvider } from './implementations/ReactToastifyProvider';
