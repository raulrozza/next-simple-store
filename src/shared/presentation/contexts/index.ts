export * from './CartProvider';
export * from './ThemeProvider';
