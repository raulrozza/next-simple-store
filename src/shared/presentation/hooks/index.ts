import { trpc } from '@/shared/infra/http/trpc';

const { useQuery, useMutation } = trpc;

export { useQuery, useMutation };
