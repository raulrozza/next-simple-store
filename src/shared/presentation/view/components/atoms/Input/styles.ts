import styled, { css } from 'styled-components';

interface InputProps {
    hasErrors?: boolean;
}

export const StyledInput = styled.input<InputProps>`
    ${({ theme, hasErrors }) => css`
        background-color: ${theme.palette.gray['0']};
        border: 2px solid transparent;
        outline: none;
        border-radius: ${theme.layout.borderRadius.sm};

        width: 100%;
        padding: ${theme.layout.spacing(2)};
        transition: outline 0.2s, border 0.2s;

        &:focus {
            border-color: ${theme.palette.primary['400']};
        }

        ${hasErrors &&
        css`
            border-color: ${theme.palette.error['400']};

            &:focus {
                border-color: ${theme.palette.error['200']};
            }
        `}
    `}
`;

export const ErrorContainer = styled.span`
    ${({ theme }) => css`
        color: ${theme.palette.error['400']};
        font-size: ${theme.typography.sizes.sm};
        padding: ${theme.layout.spacing(1, 1, 0, 1)};
    `}
`;
