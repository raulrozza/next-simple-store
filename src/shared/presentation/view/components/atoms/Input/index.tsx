import { FC } from 'react';

import { useField } from 'formik';

import { IInput } from '@/shared/presentation/entities/Field';

import { ErrorContainer, StyledInput } from './styles';

const Input: FC<IInput> = ({ name, ...props }) => {
    const [field, meta] = useField(name);

    const hasErrors = Boolean(meta.touched && meta.error);

    return (
        <div>
            <StyledInput {...props} {...field} hasErrors={hasErrors} />

            {hasErrors && <ErrorContainer>{meta.error}</ErrorContainer>}
        </div>
    );
};

export default Input;
