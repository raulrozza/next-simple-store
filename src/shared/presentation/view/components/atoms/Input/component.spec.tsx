import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Formik } from 'formik';

import { ThemeProvider } from '@/shared/presentation/contexts';

import Input from './index';

describe('Input', () => {
    it('should render the input correctly', () => {
        const view = render(
            <Formik
                initialValues={{
                    input: '',
                }}
                onSubmit={jest.fn()}
            >
                <Input name="input" placeholder="My input" />
            </Formik>,
            {
                wrapper: ThemeProvider,
            },
        );

        expect(screen.getByPlaceholderText('My input')).toBeInTheDocument();
        expect(screen.getByPlaceholderText('My input')).toHaveValue('');
        expect(view).toMatchSnapshot();
    });

    describe('Value handling', () => {
        it('should init the input with the initial field value', () => {
            const view = render(
                <Formik
                    initialValues={{ input: 'my value' }}
                    onSubmit={jest.fn()}
                >
                    <Input name="input" />
                </Formik>,
                { wrapper: ThemeProvider },
            );

            expect(screen.getByDisplayValue(/my value/)).toHaveValue(
                'my value',
            );
            expect(view).toMatchSnapshot();
        });

        it('should change the field value when typing the input', async () => {
            render(
                <Formik
                    initialValues={{ input: 'my value' }}
                    onSubmit={jest.fn()}
                >
                    <Input name="input" />
                </Formik>,
                { wrapper: ThemeProvider },
            );

            const input = screen.getByDisplayValue(/my value/);
            await userEvent.type(input, 'new value');

            expect(screen.getByDisplayValue(/new value/)).toBeDefined();
        });

        xit('should adapt the field value to mask if was passed', async () => {
            render(
                <Formik
                    initialValues={{ input: '12345678910' }}
                    onSubmit={jest.fn()}
                >
                    <Input name="input" mask="cpf" />
                </Formik>,
                { wrapper: ThemeProvider },
            );

            expect(screen.getByDisplayValue('123.456.789-10')).toBeDefined();
        });
    });

    describe('Error handling', () => {
        it('should show the error message if the input has errors and has been touched', async () => {
            const user = userEvent.setup();

            render(
                <Formik
                    initialValues={{
                        input: '',
                    }}
                    validate={() => ({ input: 'My error' })}
                    onSubmit={jest.fn()}
                    validateOnChange
                >
                    <Input name="input" placeholder="My input" />
                </Formik>,
                {
                    wrapper: ThemeProvider,
                },
            );

            expect(screen.queryByText('My error')).not.toBeInTheDocument();

            await user.type(
                screen.getByPlaceholderText('My input'),
                'new value',
            );
            act(() => {
                screen.getByPlaceholderText('My input').blur(); // Deselect to trigger touch
            });

            expect(await screen.findByText('My error')).toBeInTheDocument();
        });

        it('should not display the error message if the input was not touched', async () => {
            render(
                <Formik
                    initialValues={{ input: 'my value' }}
                    validate={() => ({ input: 'my error' })}
                    onSubmit={jest.fn()}
                    validateOnChange
                >
                    <Input name="input" placeholder="My input" />
                </Formik>,
                { wrapper: ThemeProvider },
            );

            await userEvent.type(
                screen.getByPlaceholderText('My input'),
                'invalid text',
            );

            expect(screen.queryByText(/my error/)).toBeNull();
        });
    });
});
