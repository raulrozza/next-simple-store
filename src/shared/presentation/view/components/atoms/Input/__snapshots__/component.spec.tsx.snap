// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Value handling should init the input with the initial field value 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  background-color: #FFFFFF;
  border: 2px solid transparent;
  outline: none;
  border-radius: 0.25rem;
  width: 100%;
  padding: 0.5rem;
  -webkit-transition: outline 0.2s,border 0.2s;
  transition: outline 0.2s,border 0.2s;
}

.c0:focus {
  border-color: #8CCFDE;
}

<body>
    <div>
      <div>
        <input
          class="c0"
          name="input"
          value="my value"
        />
      </div>
    </div>
  </body>,
  "container": .c0 {
  background-color: #FFFFFF;
  border: 2px solid transparent;
  outline: none;
  border-radius: 0.25rem;
  width: 100%;
  padding: 0.5rem;
  -webkit-transition: outline 0.2s,border 0.2s;
  transition: outline 0.2s,border 0.2s;
}

.c0:focus {
  border-color: #8CCFDE;
}

<div>
    <div>
      <input
        class="c0"
        name="input"
        value="my value"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Input should render the input correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  background-color: #FFFFFF;
  border: 2px solid transparent;
  outline: none;
  border-radius: 0.25rem;
  width: 100%;
  padding: 0.5rem;
  -webkit-transition: outline 0.2s,border 0.2s;
  transition: outline 0.2s,border 0.2s;
}

.c0:focus {
  border-color: #8CCFDE;
}

<body>
    <div>
      <div>
        <input
          class="c0"
          name="input"
          placeholder="My input"
          value=""
        />
      </div>
    </div>
  </body>,
  "container": .c0 {
  background-color: #FFFFFF;
  border: 2px solid transparent;
  outline: none;
  border-radius: 0.25rem;
  width: 100%;
  padding: 0.5rem;
  -webkit-transition: outline 0.2s,border 0.2s;
  transition: outline 0.2s,border 0.2s;
}

.c0:focus {
  border-color: #8CCFDE;
}

<div>
    <div>
      <input
        class="c0"
        name="input"
        placeholder="My input"
        value=""
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
